### DISCUSSION TITLE
Result difference between android.opengl.Matrix & com.google.android.filament.utils.Matrix

### DISCUSSION THREAD

**Discussion by Qheb** [google/filament#6778] [Q&A] [answered]
Hello,

While struggling with some transformations in Filament, I tried the following code :

```kotlin
val mat4 = translation(Float3(1f, 2f, 3f))
Log.i("mat4", mat4.toFloatArray().contentToString())

val matrix = FloatArray(16)
Matrix.setIdentityM(matrix, 0)
Matrix.translateM(matrix, 0, 1f, 2f, 3f)
Log.i("matrix", matrix.contentToString())
```

Since, from what I understand, both APIs are handling matrices in column-major order, I would have expected the same result for `mat4` and `matrix`. But it turns out it's not :

```
mat4      [1.0, 0.0, 0.0, 1.0, 0.0, 1.0, 0.0, 2.0, 0.0, 0.0, 1.0, 3.0, 0.0, 0.0, 0.0, 1.0]
matrix    [1.0, 0.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 0.0, 1.0, 0.0, 1.0, 2.0, 3.0, 1.0]
```

Could someone please indicate me what I am missing ? :)
Thanks !

**Comment by romainguy** [comment, marked as answer]
Both matrices are in column-major order in how they store the data. However when calling `toFloatArray()`, Filament's `Mat4` (which is from [kotlin-math](https://github.com/romainguy/kotlin-math/blob/main/src/commonMain/kotlin/dev/romainguy/kotlin/math/Matrix.kt)) returns data in a row-major format.

**Comment by Qheb** [comment]
Oh, thanks @romainguy , that explains quite a lot.

Is this the same for the `Mat4.of()` constructor method, does it take a row-major array as input ?

This means that there is an error in the [article I used for getting started with Filament](https://medium.com/@philiprideout/getting-started-with-filament-on-android-d10b16f0ec67) :grimacing: 

```kotlin
private fun Int.getTransform(): Mat4 {
   val tm = modelViewer.engine.transformManager
   return Mat4.of(*tm.getTransform(tm.getInstance(this), null))
}

private fun Int.setTransform(mat: Mat4) {
   val tm = modelViewer.engine.transformManager
   tm.setTransform(tm.getInstance(this), mat.toFloatArray())
}
```

**Reply by romainguy** [reply]
of() is row-major as well so you can write the matrix the "natural" way in your code:

```
of(
  column1_x, column2_x
  column1_y, column2_x
)
```

And yes, the article is incorrect. We could also change `toFloatArray()` to return column-major… Or provide a second variant for that.

**Comment by Qheb** [comment]
Ok, thanks for the clarifications !

In my opinion, yes a second variant might help understanding the specific aspects of the methods.
Even also another constructor method, for example `fromArray(columnMajorArray: FloatArray)`, could indicate that `of()` is also row-major (even if after seeing the example of writing a matrix the "natural" way, it makes sense :slightly_smiling_face: ).
